### PR TITLE
contrib/init: unshare mount namespace for inits

### PR DESCRIPTION
**Pull request by vbatts**
- openrc
- sysvinit-debian
- upstart

Signed-off-by: Vincent Batts vbatts@redhat.com

### CODE REVIEW

**Comment by vbatts** [comment]
ping @tianon

File: contrib/init/sysvinit-debian/docker [hunk 1]
```diff
@@ -30,6 +30,7 @@ DOCKER_SSD_PIDFILE=/var/run/$BASE-ssd.pid
 DOCKER_LOGFILE=/var/log/$BASE.log
 DOCKER_OPTS=
 DOCKER_DESC="Docker"
+UNSHARE_BINARY=${UNSHARE_BINARY:-/usr/bin/unshare}
```
Comment (tianon) [inline contrib/init/sysvinit-debian/docker]: This one should probably just be `UNSHARE` to match `DOCKER` (where the openrc one actually has `DOCKER_BINARY` so `UNSHARE_BINARY` there is consistent).

File: contrib/init/upstart/docker.conf [hunk 2]
```diff
@@ -37,7 +37,7 @@ script
 	if [ -f /etc/default/$UPSTART_JOB ]; then
 		. /etc/default/$UPSTART_JOB
 	fi
-	exec "$DOCKER" -d $DOCKER_OPTS
+	exec unshare -m -- "$DOCKER" -d $DOCKER_OPTS
```
Comment (tianon) [inline contrib/init/upstart/docker.conf]: Why the fully explicit `--mount` on the others, but only `-m` on this one?

Comment (vbatts) [inline contrib/init/upstart/docker.conf]: can i just say "because" ?
;-)

Comment (vbatts) [inline contrib/init/upstart/docker.conf]: As discussed on IRC, no difference, just a short and long version of the
same flag,give the space on the line.

File: contrib/init/sysvinit-debian/docker (hunk 1 above)
Comment (tianon) [inline contrib/init/sysvinit-debian/docker]: So you're going to swap this one, right? :innocent:

Comment (vbatts) [inline contrib/init/sysvinit-debian/docker]: done

**Comment by tianon** [comment]
Neato, LGTM; now we "just" need to test and make sure nothing breaks. :+1:

**Comment by tianon** [comment]
Just tested the OpenRC change; Docker daemon started up just fine and brought up my `--restart=always` containers.  Anything more than that we ought to test and verify, @vbatts?

**Comment by vbatts** [comment]
@tianon restart it, such that it does those init targets for the service fine. 

and `sudo readlink /proc/1/ns/mnt /proc/$(pgrep docker| head -1)/ns/mnt`

**Comment by vbatts** [comment]
(the readlink should show different numbers)

**Comment by tianon** [comment]
Cool, all that's working great.  One down, two to go!

**Comment by jfrazelle** [comment]
will test today :)

**Comment by jfrazelle** [comment]
these are not folder on debian wheezy when testing sysvinit :( /proc/1/ns/mnt /proc/$(pgrep docker| head -1)/ns/mnt

**Comment by jfrazelle** [comment]
but everything restarts and works

**Comment by jfrazelle** [comment]
testing ubuntu now

**Comment by jfrazelle** [comment]
works on ubuntu, so just sysvinit is the prob :)

**Comment by tianon** [comment]
Even with the proper kernel from wheezy-backports?

**Comment by jfrazelle** [comment]
:angel: whoops

**Comment by jfrazelle** [comment]
perfect works!

**Comment by jfrazelle** [comment]
LGTM

**Comment by jfrazelle** [comment]
literally the most intense thing to test

**Comment by tianon** [comment]
:balloon: :cake:

**Comment by paulmenzel** [comment]
Hi. Looking through changes of `contrib/init/upstart/docker.conf`, I am wondering about the reason for this change. Why should the mount namespace be unshared. What problems does this fix?

**Comment by vbatts** [comment]
@paulmenzel most notably the storage-driver `devicemapper` creates a new device on every container start, and destroys it on container stop/removal. When docker is on the host's mount namespace, the during the time the device exist, if another PID on the host unshares mount namespace (e.g. the cups daemon), that PID will include reference to the device in its mountinfo (even though that PID doesn't care about the device), then docker fails to stop/remove the container.

Similarly, if any workflow _inside_ containers involves creating or mounting, then this issue will cause EBUSY (device or resource busy) as well.

The only way to cleanly fix this is unshare the mount namespace.

**Comment by paulmenzel** [comment]
@vbatts, thank you for the clarification. It’d be awesome to have such things in the commit message the next time.

Thank you again for your work!

**Pull request merged**